File: src/main/kotlin/hyunsoo/65week/괄호 회전하기.kt
package hyunsoo.`65week`

import java.util.*
 /**
 *
 * <문제>
 * [괄호 회전하기](https://school.programmers.co.kr/learn/courses/30/lessons/76502)
 *
 * - 아이디어
 *
 * - 트러블 슈팅
 *
 */
class `전현수_괄호_희전하기` {

    private val opens = listOf('(', '[', '{')

    fun solution(s: String): Int {
        var answer = 0

        val target = s + s

        var start = 0
        repeat(s.length) {
            if (target.slice(start until start + s.length).isAlright()) answer++
            start++
        }

        return answer
    }

     private fun String.isAlright(): Boolean {

         val stack = Stack<Char>()

         this.forEachIndexed { index, ch ->

             if (ch in opens) {
                 stack.add(ch)
             }
             else {

                 if (stack.isEmpty()) return false

                 when (stack.peek()) {
                     '(' -> {
                         if (ch == ')') stack.pop()
                         else return false
                     }
                     '[' -> {
                         if (ch == ']') stack.pop()
                         else return false
                     }
                     '{' -> {
                         if (ch == '}') stack.pop()
                         else return false
                     }
                 }
             }
         }

         return stack.isEmpty()

     }
}

fun main() {
    전현수_괄호_희전하기().solution("[](){}")
}

File: src/main/kotlin/jimin/65week/괄호 회전하기.py
def solution(s):
    answer = 0
    s = list(s)
    for i in range(len(s)):
        if isCorrect(s):
            answer += 1
        tmp = s[0]
        s.pop(0)
        s.append(tmp)

    return answer


def isCorrect(s):
    stack = []
    idx = 0
    isCorrect = True
    start = ['(', '[', '{']
    end = [')', ']', '}']
    while idx < len(s) and isCorrect:
        if s[idx] in start:
            stack.append(s[idx])
        elif s[idx] in end:
            if not stack:
                isCorrect = False
            else:
                for i in range(3):
                    if stack[-1] == start[i]:
                        if s[idx] == end[i]:
                            stack.pop()
                        else:
                            isCorrect = False
                        break
        idx += 1

    if stack:
        isCorrect = False

    return isCorrect